src/test/java/org/opensearch/security/dlic/rest/api/ConfigUpgradeApiActionUnitTest.java

```diff
@@ -0,0 +1,291 @@
+        // Setup
+        doThrow(new IOException("abc")).when(configUpgradeApiAction).loadConfigFileAsJson(any());
+
+        // Execute
+        configUpgradeApiAction.canUpgrade(restChannel, restRequest, client);
+
+        // Assert
+        verify(restChannel).sendResponse(verifyResponseBody(body -> assertThat(body, containsString("see the log file to troubleshoot"))));
+    }
+
+    @Test
+    public void testPerformUpgrade_ErrorLoadingConfig() throws Exception {
+        // Setup
+        doThrow(new IOException("abc")).when(configUpgradeApiAction).loadConfigFileAsJson(any());
+
+        // Execute
+        configUpgradeApiAction.performUpgrade(restChannel, restRequest, client);
+
+        // Assert
+        verify(restChannel).sendResponse(verifyResponseBody(body -> assertThat(body, containsString("see the log file to troubleshoot"))));
+    }
+
+    @Test
+    public void testPerformUpgrade_ErrorApplyConfig() throws Exception {
+        // Setup
+        doThrow(new RuntimeException("abc")).when(configUpgradeApiAction).patchEntities(any(), any(), any());
+
+        // Execute
+        configUpgradeApiAction.performUpgrade(restChannel, restRequest, client);
+
+        // Assert
+        verify(restChannel).sendResponse(verifyResponseBody(body -> assertThat(body, containsString("see the log file to troubleshoot"))));
+    }
+
+    @Test
+    public void testPerformUpgrade_NoDifferences() throws Exception {
+        // Setup
+        final var rolesCopy = rolesConfiguration.deepClone();
+        rolesCopy.removeStatic(); // Statics are added by code, not by config files, they should be omitted
+        final var rolesJsonNode = Utils.convertJsonToJackson(rolesCopy, true);
+        doReturn(rolesJsonNode).when(configUpgradeApiAction).loadConfigFileAsJson(any());
+
+        // Execute
+        configUpgradeApiAction.performUpgrade(restChannel, restRequest, client);
+
+        // Verify
+        verify(restChannel).sendResponse(verifyResponseBody(body -> assertThat(body, containsString("no differences found"))));
+    }
+
+    @Test
+    public void testPerformUpgrade_WithDifferences() throws Exception {
+        // Execute
+        configUpgradeApiAction.performUpgrade(restChannel, restRequest, client);
+
+        // Verify
+        verify(restChannel).sendResponse(argThat(response -> {
+            final var rawResponseBody = response.content().utf8ToString();
+            final var newlineNormalizedBody = rawResponseBody.replace("\r\n", "\n");
+            assertThat(newlineNormalizedBody, equalTo("{\n" + //
+                "  \"status\" : \"OK\",\n" + //
+                "  \"upgrades\" : {\n" + //
+                "    \"roles\" : {\n" + //
+                "      \"add\" : [ \"new_role\" ]\n" + //
+                "    }\n" + //
+                "  }\n" + //
+                "}"));
+            return true;
+        }));
+    }
+
+    @Test
+    public void testConfigurationDifferences_OperationBash() throws IOException {
```

Comment:
Does "OperationBash" mean something in this context? I don't see any connection to shells below unless I am missing something.

Reply:
I meant bash as in a party or large gathering. as this test is is following a very similar pattern to setup/act/verify different conditions of the differences produced.  Do you have an alternative name suggestion that would work better?

Reply:
I'm going to keep this as is, happy to make a small follow up PR for improvements in this space